sound/soc/sof/mediatek/mt8195/mt8195.c

```diff
@@ -352,6 +379,10 @@ const struct snd_sof_dsp_ops sof_mt8195_ops = {
 	/* misc */
 	.get_bar_index	= mt8195_get_bar_index,
 
+	/* machine driver */
+	.machine_select = mt8195_machine_select,
+	.machine_register = sof_machine_register,
+
```

Comment:
Sorry @yaochunhung I am really confused. In previous PRs you described how the same card can include SOF and non-SOF managed streams. But above you seem to describe a case where the card is created in two different ways.

I don't get how this might work.

Reply:
@plbossart Sorry for confusion. For sound card with SOF support, you can just reference the binding below and ignore item(2). Where item(2) is used for  sound card without SOF support. 
```
item(3)
  adsp: adsp@10803000 {
       compatible =  "mediatek,mt8195-dsp";
...
       sound {
              mediatek,dptx-codec = <&dp_tx>;
              mediatek,hdmi-codec = <&hdmi0>;
              mediatek,platform = <&afe>;
             };
```
For item(3), we define compatible =  "mediatek,mt8195-dsp" in device tree for sof driver and use mt8195_machine_select/sof_machine_register to select and register machine driver. 
We define two machine driver files with the same driver name "mt8195_mt6359_rt1019_rt5682" for with SOF support and without SOF support.
With SOF support(for item3) : sound/soc/mediatek/mt8195/mt8195-mt6359-rt1019-rt5682-sof.c
https://patchwork.kernel.org/project/linux-mediatek/patch/20211103100040.11933-5-trevor.wu@mediatek.com/
Without SOF support(for item2) : ound/soc/mediatek/mt8195/mt8195-mt6359-rt1019-rt5682.c
https://patchwork.kernel.org/project/linux-mediatek/patch/20211103100040.11933-4-trevor.wu@mediatek.com/
Please feel free to tell me if my explanation is still not clear to you. Thanks.

Reply:
I am not able to follow any of the explanations, sorry. Someone with a better understanding of device tree will have to review this. It seems Mark Brown is equally confused btw.

Reply:
I think this is kind of backwards...
The sound card node (compatible = "mediatek,mt8195_mt6359_rt1019_rt5682") should have phandle to the DSP node.
The driver for the DSP should be registering the FE DAIs and platform component for the DSP.

The DSP driver is not a sound card driver, the paltfrom_driver of DSP should avoid creating child platform_device as in principle in DT systems platform_devices are unique SoC or board level devices.

Your non SOF driver uses DAIs (phandle: `mediatek,platform`) DMAs (ASoC platform driver, likely dmaengine-pcm, attached to the DAI's platform_device) and codecs.
When you add SOF support, you add to these the DSP, it is going to create new PCM or take over paths from the non DSP use.

The compatible is a tricky question as the hardware has not changed, so in theory you need to keep the string and decide to use SOF if the DSP node is present and build up the routes accordingly.
It might be allowed to have different compatible for DSP enabled audio support, but it can be avoided with carefully selected 'markers' in DT to handle the two world.

Reply:
@ujfalusi  Thanks for your detail descriptions. I will base on your suggestion to modify the DSP node as property of sound card node as below,
``` 
&sound {
	pinctrl-names = "default";
	pinctrl-0 = <&aud_pins_default>;
	mediatek,adsp = <&adsp>;
        mediatek,platform =<&afe>;
	status = "okay";
}
```
where adsp is DSP node for dsp driver and afe is regular(non-sof) node for regular pl. Therefore, machine register won't be needed in this patch. But I still keep the machine_select to get sof_tplg_filename. Please feel free to tell me if I am misunderstanding. Thanks.

Reply:
@ujfalusi @plbossart Thanks for review. The machine driver is applied to mark/for-next branch as the link https://git.kernel.org/pub/scm/linux/kernel/git/broonie/sound.git/commit/?id=3d00d2c07f04f47aa4228700b440ac47abf13853. Please help to check if you have any concern or suggestions about this PR. Thanks.

Reply:
I agree with @dbaluta and @plbossart that the `machine_select` should not be used, but you should be using the of match data in a  similar way iMX is doing.

Reply:
@dbaluta  I check this function sof_machine_check and snd_sof_machine_select should return mach to avoid fail. Could you please advice how to avoid it without providing machine_select callback on iMX? Otherwise, I think sof_tplg_filename is also needed, but I don't know how to assign it without call machine_select . Thanks.

Reply:
@yaochunhung I think this might be useful. https://www.spinics.net/lists/devicetree/msg433908.html

But I hit a roadblock and didn't get a chance to find a better solution.

I think for now at least we can have a default_tplg_filename in sof_dev_desc and select it based on probed platform.